### PR TITLE
IETF 116:  Update Subscribe Messages

### PR DESCRIPTION
**Pull request by suhasHere**
This PR depends on #121 and has the following changes  to capture IETF 116 decisions 

1. Rename subscribe message to subscribe request 
2. Include track names and authorization in the subscribe request
3. Add  basic support for subscribe reply providing the results of subscription
4. Add support for short compression identifiers as Track Id (varint) to be used in the object header messages

Currently the subscribe granularity is per track and a future PR can be modified to support multiple tracks if needed.

### CODE REVIEW

File: draft-lcurley-warp.md [hunk 1]
```diff
@@ -603,7 +605,7 @@ OBJECT Message {
 {: #warp-object-format title="Warp OBJECT Message"}
 
 * Track ID:
-The track identifier as declared in CATALOG ({{message-catalog}}).
+The track identifier obtained as part of subscription and/or publish control message exchanges.
```
Comment (afrind) [inline draft-lcurley-warp.md:608]: I think it would be good to explain Track Name vs Track ID with some expository text where tracks are defined (maybe in #121).

Then here perhaps you can say "Track ID: unique identifier for this track within the current (session? connection?)"

File: draft-lcurley-warp.md [hunk 2]
```diff
@@ -664,29 +666,116 @@ This contains base information required to decode OBJECT messages, such as codec
 An endpoint MUST NOT send multiple CATALOG messages.
 A future draft will add the ability to add/remove/update tracks.
 
-## SUBSCRIBE {#message-subscribe}
-After receiving a CATALOG message ({{message-catalog}}, the receiver sends a SUBSCRIBE message to indicate that it wishes to receive the indicated tracks.
+## SUBSCRIBE REQUEST {#message-subscribe-req}
+
+Entities that intend to receive media will do so via subscriptions to
+one or more tracks. The information for the tracks can be obtained via catalog or from incoming SUBSCRIBE messages.
```
Comment (afrind) [inline draft-lcurley-warp.md:672]: Did you mean SUBSCRIBE_REQUEST?  SUBSCRIBE is no longer defined?

File: draft-lcurley-warp.md [hunk 3]
```diff
@@ -664,29 +666,116 @@ This contains base information required to decode OBJECT messages, such as codec
...
+  Track Authorization Info (...)...
+  [Group Sequence (i)],
+  [Object Sequence (i)],
+}
+~~~
+{: #warp-track-info-format title="Warp TRACK INFO Content"}
+
+
+* Track Name:
+Identifies the fully qualified track name as defined in ({{model-track}}).
```
Comment (afrind) [inline draft-lcurley-warp.md:703]: Does it always have to be fully qualified, or can that vary by "Streaming format" or another profile.  My understanding is that some use cases would like to only send the suffix here, and combine it with a prefix communicated elsewhere.  Is that possible?

File: draft-lcurley-warp.md [hunk 4]
```diff
@@ -664,29 +666,116 @@ This contains base information required to decode OBJECT messages, such as codec
...
+
+* Track ID:
```
Comment (afrind) [inline draft-lcurley-warp.md:705]: Is it better to have the publisher assign the track ID in the SUBSCRIBE_REPLY or the subscriber in the SUBSCRIBE_REQUEST?  Are there any tradeoffs?

**Review by fluffy — APPROVED** [review]
This seems about right.

File: draft-lcurley-warp.md [hunk 5]
```diff
@@ -664,29 +666,116 @@ This contains base information required to decode OBJECT messages, such as codec
...
+A proposal serving as short hop-by-hop identifier to be used in the OBJECT ({{message-object}}) message header for the requested tack. Given that the media corresponding to a track can potentially be delivered over multiple data streams, the `Track ID` provides the necessary mapping between the "Track Name" in the control message and the corresponding data streams. It also serves as a compression identifier for containing the size of object headers instead of carrying complete "Track Name" information in every object message. A client MUST choose an `Track ID` value that is unique within a QUIC Connection.
+
+* Group Sequence:
+Identifies the group within the track to start delivering 
+the objects. This field is optional and when omitted, the
+publisher MUST deliver the objects from the most recent 
+group.
+
+* Object Sequence:
```
Comment (afrind) [inline draft-lcurley-warp.md:714]: I don't think the layout can have both Group and Object sequence be optional -- if there's only one varint present in the message it's implicitly a Group Sequence, right?  

I read your intent to support these three cases:

1. No group/object info - send all groups/objects starting from beginning of most recent group, object 0
2. Group sequence - send from this specific group, starting from object 0
3. Group sequence and Object sequence - send only this specific object from this specific group

Is that right?

File: draft-lcurley-warp.md [hunk 6]
```diff
@@ -664,29 +666,116 @@ This contains base information required to decode OBJECT messages, such as codec
...
+beginning of the selected group, i.e., starting at Object Sequence of 0.
+
+* Track Authorization Info: 
+Carries track authorization, typically in the form of a token, authorizing client’s subscription to the track. The specifics of obtaining the authorization information is out of scope for this specification.
+
+Subscriptions stay active until one of the following happens:
+
+- a client local policy dictates expiration of a subscription.
+- optionally, a server policy dicates subscription expiration.
+- the underlying transport is disconnected.
```
Comment (afrind) [inline draft-lcurley-warp.md:728]: If we want automatic unsubscribe, then I think SETUP or SUBSCRIBE_REPLY needs to exchange information about the publisher's required refresh period.

I also think we want an explicit UNSUBSCRIBE message, in case the publishers refresh interval is much longer than the receiver wants to continue getting this track.

File: draft-lcurley-warp.md [hunk 7]
```diff
@@ -664,29 +666,116 @@ This contains base information required to decode OBJECT messages, such as codec
...
+{
+  TRACK RESPONSE Track
+}
+~~~
+{: #warp-subscribe-reply format title="Warp SUBSCRIBE REPLY Message"}
+
+* Track:
+Captures the result of subscription as defined in `TRACK RESPONSE`
+
+The `TRACK RESPONSE` structure is as defined below
```
Comment (afrind) [inline draft-lcurley-warp.md:747]: Is TRACK RESPONSE going to be reused elsewhere (maybe PUBLISH_REQUEST/RESPONSE)?  If not, just merge the fields there into SUBSCRIBE_REPLY?

File: draft-lcurley-warp.md [hunk 8]
```diff
@@ -664,29 +666,116 @@ This contains base information required to decode OBJECT messages, such as codec
...
+
+~~~
+enum Track Result
+{
+  Ok(0),
+  Expired(1),
```
Comment (afrind) [inline draft-lcurley-warp.md:753]: It's too early to bikeshed error codes, but I'm curious if you called out Expired here separately from Fail so that a publisher can send a SUBSCRIBE_REPLY once the track expires?  Eg, are you envisioning a sequence like:

Consumer -> Publisher: SUBSCRIBE_REQUEST
P -> C: SUBSCRIBE_REPLY
P -> C: OBJECT*
...
P -> C: SUBSCRIBE_REPLY (Expired)

File: draft-lcurley-warp.md [hunk 9]
```diff
@@ -664,29 +666,116 @@ This contains base information required to decode OBJECT messages, such as codec
...
+
+* Track Name:
+Identifies the track in the request message for which this
+response is provided.
+
+* Response:
+Provides result of the transaction, where a value of `Ok` indicates successful subscription. For failed or expired responses, the "Reason Phrase" shall be populated with appropriate reason code.
+
+* Track ID:
+Identifies the hop-by-hop identifier mapping the given Track Name to be populated in the Object messages ({{message-object}}). This field is optional and is provided in cases where the end point is unable to use the `Track ID` proposed by the peer in the request message. If populated, the `Track ID` field in the Object's header messages MUST be populated with the value in this field otherwise the `Track ID` value MUST correspond to one in the request message for a given track.
```
Comment (afrind) [inline draft-lcurley-warp.md:777]: If the publisher has full control over the ID space I think I would recommend against letting the client ask for particular IDs.  Do you have a use case in mind or would publisher-assigned work?

File: draft-lcurley-warp.md [hunk 10]
```diff
@@ -664,29 +666,116 @@ This contains base information required to decode OBJECT messages, such as codec
...
+~~~
+{: #warp-track-response-format title="Warp TRACK RESPONSE Content"}
+
+
+* Track Name:
+Identifies the track in the request message for which this
+response is provided.
+
+* Response:
+Provides result of the transaction, where a value of `Ok` indicates successful subscription. For failed or expired responses, the "Reason Phrase" shall be populated with appropriate reason code.
```
Comment (afrind) [inline draft-lcurley-warp.md:774]: I think shall is a normative verb, since it's optional, you can use MAY or can.  Also maybe say "subscription request" rather than "transaction".

**Review by afrind — COMMENTED** [review]
Individual review

File: draft-lcurley-warp.md (hunk 2 above)
Comment (suhasHere) [reply draft-lcurley-warp.md:672]: good catch, will update

File: draft-lcurley-warp.md (hunk 3 above)
Comment (suhasHere) [reply draft-lcurley-warp.md:703]: We discussed this topic in length at the IETF116 in the context of WT and Native QUIC and why having full name will work for both cases uniformly and also for inter distribution network flows.

We concluded that  tracks are central component and they are self identifying and are carry that way in subscribes/publishes.

**Comment by suhasHere** [comment]
@afrind 

Good points Alan on the comments. I will respond and update text to reflect these by early next week.

File: draft-lcurley-warp.md (hunk 3 above)
Comment (afrind) [reply draft-lcurley-warp.md:703]: @vasilvv @kixelated will this meet your use cases?  My understanding of Mo's proposal on the mailing list was specifically so that suffix only *could* be specified here -- both suffix only or fully qualified are OK.

File: draft-lcurley-warp.md (hunk 1 above)
Comment (afrind) [reply draft-lcurley-warp.md:608]: I'm not necessarily hung up on which PR defines it - only the name vs id concept deserves top level mention earlier in the document.

File: draft-lcurley-warp.md (hunk 7 above)
Comment (suhasHere) [reply draft-lcurley-warp.md:747]: yes it is used across Publish and Subscribe message .. same with Track Info too

File: draft-lcurley-warp.md (hunk 3 above)
Comment (suhasHere) [reply draft-lcurley-warp.md:703]: I think we did have a discussion at the mic on this aspect and we agreed to keep the behavior of subscribe and publish unifirm across all the nodes, it is best to keep the track as self identifying without tying it to the connection context.  Since this will allow 2 things
1.  Work in the same way across WT and native QUIC. For WT there is extra bytes of duplication which is not concerning
2. Nodes processing subscribe/publish can be uniformly defined without needing to explcitly specify the cases where one option would work vs the other. 
3. Tracks are authorized and one can define security properties in a common way across applications and distribution networks

File: draft-lcurley-warp.md (hunk 4 above)
Comment (suhasHere) [reply draft-lcurley-warp.md:705]: Good question @afrind 

In the current model, the publisher has the final say on the trackId to be used. 

The benefit of Subscriber Request proposing the TrackId is it will lead to a design where one can minimize the rewrites of the object headers per hop if there is no conflict. 

Again the publisher is the one that can decide the trackId and it will do so in the SUBSCRIBE_REPLY.

File: draft-lcurley-warp.md (hunk 6 above)
Comment (suhasHere) [reply draft-lcurley-warp.md:728]: @afrind  great points. I agree on both the points. I will add issues to cover those.

File: draft-lcurley-warp.md (hunk 10 above)
Comment (suhasHere) [reply draft-lcurley-warp.md:774]: +1 .. will update

File: draft-lcurley-warp.md (hunk 9 above)
Comment (suhasHere) [reply draft-lcurley-warp.md:777]: the only benefit is minimize object header rewrites. May be very are optimizing too early. i am fine with making it publisher controlled id space.

File: draft-lcurley-warp.md (hunk 6 above)
Comment (suhasHere) [reply draft-lcurley-warp.md:699]: #133 , #132  created these issues

File: draft-lcurley-warp.md (hunk 8 above)
Comment (suhasHere) [reply draft-lcurley-warp.md]: Removed expired for now. We can revisit later if its needed.

**Comment by suhasHere** [comment]
@afrind pushed few changes to address your review feedback. please let me know thanks

File: draft-lcurley-warp.md [hunk 11]
```diff
@@ -169,6 +169,11 @@ Track Name:
 
 : See {{track-name}}.
 
+Track ID:
```
Comment (kixelated) [inline draft-lcurley-warp.md:172]: I don't think this should be in the definitions. I think Alan was asking for a section on how name/URL maps to ID.

File: draft-lcurley-warp.md [hunk 12]
```diff
@@ -650,29 +657,97 @@ This contains base information required to decode OBJECT messages, such as codec
...
+Subscriptions stay active until one of the following happens:
+
+- a client local policy dictates expiration of a subscription.
+- optionally, a server policy dicates subscription expiration.
+- the underlying transport is disconnected.
+
+On successful subscription, the publisher MUST deliver objects
+from the beginnig of the most recent group in the track, i.e, starting at Object Sequence of 0.
+
+A client MUST renew its subscriptions by sending a new `SUBSCRIBE` to keep them active. Such subscriptions MUST refresh the existing subscriptions and thus only the most recent SUBSCRIBE message is considered active. A renewal period of 5 seconds is RECOMMENDED.
```
Comment (kixelated) [inline draft-lcurley-warp.md:704]: This is a hard no from me as discussed in #102 .

I would be okay with the server indicating a subscription expiration in the SUBSCRIBE_REPLY message. That way the consumer (nit: not client) would resubscribe based on the policy set by the publisher. 

I would still want to talk about the use-case though; what problem are you solving?

File: draft-lcurley-warp.md [hunk 13]
```diff
@@ -650,29 +657,97 @@ This contains base information required to decode OBJECT messages, such as codec
...
+* Track Authorization Info: 
+Carries track authorization, typically in the form of a token, authorizing the client’s subscription to the track. The specifics of obtaining the authorization information is out of scope for this specification.
+
+Subscriptions stay active until one of the following happens:
+
+- a client local policy dictates expiration of a subscription.
+- optionally, a server policy dicates subscription expiration.
+- the underlying transport is disconnected.
+
+On successful subscription, the publisher MUST deliver objects
```
Comment (kixelated) [inline draft-lcurley-warp.md:701]: The group wouldn't start at 0 when joining a live stream. The relay should not rewrite object headers either, as it will break any form of history (ex. catalog that indicates group 45 = timestamp 12:34)

File: draft-lcurley-warp.md (hunk 6 above)
Comment (kixelated) [reply draft-lcurley-warp.md:699]: I would encode an "expires" in the SUBSCRIBE_REPLY. It would be a duration (not a timestamp) so the consumer would need to refresh the subscription a little early.

File: draft-lcurley-warp.md [hunk 14]
```diff
@@ -650,29 +657,97 @@ This contains base information required to decode OBJECT messages, such as codec
...
+{
+  Ok(0),
+  Fail(1)
+}
+
+TRACK RESPONSE {
+  Track Name Length(i),
+  Track Name(...)...,
+  Track Result Response,
+  [Track ID(i)],
```
Comment (kixelated) [inline draft-lcurley-warp.md:734]: I've mentioned this in a few PRs, but these optionals are not possible to encode.

File: draft-lcurley-warp.md [hunk 15]
```diff
@@ -650,29 +657,97 @@ This contains base information required to decode OBJECT messages, such as codec
...
+enum Track Result
+{
+  Ok(0),
+  Fail(1)
+}
+
+TRACK RESPONSE {
+  Track Name Length(i),
+  Track Name(...)...,
+  Track Result Response,
```
Comment (kixelated) [inline draft-lcurley-warp.md:733]: varint?

File: draft-lcurley-warp.md (hunk 14 above)
Comment (kixelated) [reply draft-lcurley-warp.md:734]: And why is this optional in the first place?

File: draft-lcurley-warp.md [hunk 16]
```diff
@@ -650,29 +657,97 @@ This contains base information required to decode OBJECT messages, such as codec
...
+
+* Track Name:
+Identifies the track in the request message for which this
+response is provided.
+
+* Response:
+Provides result of the transaction, where a value of `Ok` indicates successful subscription. For failed response, the "Reason Phrase" is populated with appropriate reason code.
+
+* Track ID:
+A short hop-by-hop identifier to be used in the OBJECT ({{message-object}}) message header for the requested tack. Given that the media corresponding to a track can potentially be delivered over multiple data streams, the `Track ID` provides the necessary mapping between the "Track Name" in the control message and the corresponding data streams. It also serves as a compression identifier for containing the size of object headers instead of carrying complete "Track Name" information in every object message. The `Track ID` chosen MUST be unique within the QUIC Connection and is populated for successful subscriptions.
```
Comment (kixelated) [inline draft-lcurley-warp.md:750]: I think the consumer should choose the track ID in the SUBSCRIBE message.

Otherwise, there's a race condition where the SUBSCRIBE_REPLY could be delayed, while the consumer receives objects with an unknown track ID. It's a whole lot simpler if the consumer chooses the ID in the SUBSCRIBE message.

File: draft-lcurley-warp.md (hunk 11 above)
Comment (suhasHere) [reply draft-lcurley-warp.md:172]: I see .. I did have it as part of the messages section. May be @afrind  can you see the message section changes and see if it still works.

File: draft-lcurley-warp.md (hunk 13 above)
Comment (suhasHere) [reply draft-lcurley-warp.md:701]: I don't think I mentioned group starting at 0 .. I meant object starting at 0 form the most recent group

File: draft-lcurley-warp.md (hunk 14 above)
Comment (suhasHere) [reply draft-lcurley-warp.md:734]: A failed subscription doesn't need a track Id .. Options are very common QUIC encoding. Can you explain what are the challenges you see ?

File: draft-lcurley-warp.md [hunk 17]
```diff
@@ -650,29 +657,97 @@ This contains base information required to decode OBJECT messages, such as codec
...
+~~~
+{: #warp-track-response-format title="Warp TRACK RESPONSE Content"}
+
+
+* Track Name:
+Identifies the track in the request message for which this
+response is provided.
+
+* Response:
+Provides result of the transaction, where a value of `Ok` indicates successful subscription. For failed response, the "Reason Phrase" is populated with appropriate reason code.
```
[the remark this answers is not in the document]
Comment (suhasHere) [reply draft-lcurley-warp.md:747]: +1 .. will upadte

File: draft-lcurley-warp.md (hunk 11 above)
Comment (kixelated) [reply draft-lcurley-warp.md:172]: I called this "Subscription ID" in a previous PR. It might be less confusing since it's decoupled from track.

File: draft-lcurley-warp.md (hunk 13 above)
Comment (kixelated) [reply draft-lcurley-warp.md:701]: I missed the `i.e` when reading this, which should be `ex.` anyway...

I think the example is fine just make up a number like 53.

File: draft-lcurley-warp.md (hunk 14 above)
Comment (kixelated) [reply draft-lcurley-warp.md:734]: So I think I understand what you're trying to do in this instance, but you need more text. There's no magic way to make a field optional in QUIC encoding.

How about:
> The track ID field MUST ONLY be present if the result is Ok. The reason and length MUST ONLY be present if the result is not Ok.

I would just make a separate messages honestly. SUBSCRIBE_OK and SUBSCRIBE_ERROR. If we want to consolidate them in the future, we can add a flag for the type (like how QUIC frames work).

File: draft-lcurley-warp.md (hunk 11 above)
Comment (suhasHere) [reply draft-lcurley-warp.md:172]: I did call it "Subscription ID" in #96  and there was some push backs on that term IIRC . This is just naming, I am fine with either of the names. We need to pick one and move on

Comment (kixelated) [reply draft-lcurley-warp.md:172]: I don't see any push-back in that PR around the name. There was a LOT of other stuff in that PR. 

And yeah I called it "Subscribe ID" in #43. I think "Subscription ID" is better.

File: draft-lcurley-warp.md (hunk 12 above)
Comment (suhasHere) [reply draft-lcurley-warp.md:704]: i will remove it from this PR and we can discuss in #102

File: draft-lcurley-warp.md (hunk 16 above)
Comment (suhasHere) [reply draft-lcurley-warp.md:750]: i will revert some of these for a fresh discussion

**Comment by suhasHere** [comment]
@fluffy @kixelated @vasilvv @afrind please give it another read and let me know . thanks

File: draft-lcurley-warp.md (hunk 3 above)
Comment (afrind) [reply draft-lcurley-warp.md:688]: I think this discussion is captured in #121 for now and we can update based on whatever the outcome is there.

File: draft-lcurley-warp.md [hunk 18]
```diff
@@ -650,30 +652,110 @@ This contains base information required to decode OBJECT messages, such as codec
...
+}
+~~~
+{: #warp-track-info-format title="Warp TRACK INFO Content"}
+
+
+* Track Name:
+Identifies the fully qualified track name as defined in ({{model-track}}).
+
+* Track ID: 
+A proposal serving as short hop-by-hop identifier to be used in the OBJECT ({{message-object}}) message header for the requested tack. Given that the media corresponding to a track can potentially be delivered over multiple data streams, the Track ID provides the necessary mapping between the "Track Name" in the control message and the corresponding data streams. It also serves as a compression identifier for containing the size of object headers instead of carrying complete "Track Name" information in every object message. A client MUST choose an Track ID value that is unique within a QUIC Connection.
```
Comment (afrind) [inline draft-lcurley-warp.md:691]: Editorial comments:

I don't think you need to say proposal:

> An session specific identifier used in OBJECT ... message headers for the requested track.

"provides the necessary mapping between" -> maps

> ... the Track ID maps the "Track Name" ...

Restating the compression bit more succinctly:

> Track IDs are generally shorter than Track Names and thus reduce the overhead in OBJECT messages.

Comment (afrind) [inline draft-lcurley-warp.md:691]: Architectural Comment:

I'd prefer to just let the server pick the IDs rather than giving the client an option to ask for a specific one.  If we remove that functionality from this PR we can move this forward and have a separate PR to discuss the pros and cons of client hinting track IDs there.

File: draft-lcurley-warp.md [hunk 19]
```diff
@@ -650,30 +652,110 @@ This contains base information required to decode OBJECT messages, such as codec
...
+
+* Group Sequence: 
+Identifies the group within the track to start delivering the objects. This field is optional and when omitted, the publisher MUST deliver the objects from the most recent group.
+
+* Object Sequence: 
+Identifies the object within the track to start the media delivery. The `Group Sequence` field MUST be set to identify the group under which to
+the deliver the objects. This field is optional and when omitted, the publisher MUST deliver starting from the beginning of the selected group.
+
+* Track Authorization Info: 
+Carries track authorization, typically in the form of a token, authorizing the client’s subscription to the track. The specifics of obtaining the authorization information is out of scope for this specification.
```
Comment (afrind) [inline draft-lcurley-warp.md:701]: Maybe remove "typically in the form of a token", since "the specifics of ... authorization information is out of scope".

File: draft-lcurley-warp.md (hunk 6 above)
Comment (afrind) [reply draft-lcurley-warp.md:707]: I recommend you add "expires" in SUBSCRIBE_REPLY as @kixelated suggests, and rewrite this bit to say "the subscription is active until it expires, or the underlying transport is disconnected".  Once UNSUBSCRIBE is added, then you can add the bit here about "client policy".

File: draft-lcurley-warp.md (hunk 15 above)
Comment (afrind) [reply draft-lcurley-warp.md:739]: What do you want to varint here?

File: draft-lcurley-warp.md (hunk 14 above)
Comment (afrind) [reply draft-lcurley-warp.md:740]: The QUIC wg likes to use different message types for this kind of thing -- along the lines of Luke's second suggestion, but we can also move forward now with just some extra text explaining when the ID will be there and when it will not.

File: draft-lcurley-warp.md [hunk 20]
```diff
@@ -650,30 +652,110 @@ This contains base information required to decode OBJECT messages, such as codec
...
+Identifies the track in the request message for which this
+response is provided.
+
+* Response:
+Provides result of the transaction, where a value of `Ok` indicates successful subscription. For failed response, the "Reason Phrase" 
+is populated with appropriate reason.
+
+
+* Track ID:
+Identifies the hop-by-hop identifier mapping the given Track Name to be populated in the Object messages ({{message-object}}) and MUST only be present if the `Response` is `Ok`. This field is populated with either the `Track ID` value provided in the request or the one chosen by client processing the request. The Track ID field in the Object's header messages MUST be populated with the value in this field.
```
Comment (afrind) [inline draft-lcurley-warp.md:758]: > the `Track ID` value provided in the request or the one chosen by client processing the request

Do you mean "chosen by the *server* processing the request" ?

**Review by afrind — COMMENTED** [review]
Individual Review

File: draft-lcurley-warp.md [hunk 21]
```diff
@@ -650,30 +652,110 @@ This contains base information required to decode OBJECT messages, such as codec
...
+The `TRACK INFO` structure is as defined below
+
+~~~
+TRACK INFO {
+  Track Name Length(i),
+  Track Name(...),
+  Track ID (i),
+  Track Authorization Info Length(i),
+  Track Authorization Info (...),
+  [Group Sequence (i)],
```
Comment (kixelated) [inline draft-lcurley-warp.md:680]: Add some text indicating how these are optional. In the current PR, it's only possible to tell if a field is present based on the message length. This is not how QUIC, HTTP/3, QPACK, TLS, etc encodes optional fields so it needs to be very explicit.

Or make them required and reserve 0. I think we should start with that.

File: draft-lcurley-warp.md [hunk 22]
```diff
@@ -650,30 +652,110 @@ This contains base information required to decode OBJECT messages, such as codec
...
+
+* Track ID: 
+A proposal serving as short hop-by-hop identifier to be used in the OBJECT ({{message-object}}) message header for the requested tack. Given that the media corresponding to a track can potentially be delivered over multiple data streams, the Track ID provides the necessary mapping between the "Track Name" in the control message and the corresponding data streams. It also serves as a compression identifier for containing the size of object headers instead of carrying complete "Track Name" information in every object message. A client MUST choose an Track ID value that is unique within a QUIC Connection.
+
+* Group Sequence: 
+Identifies the group within the track to start delivering the objects. This field is optional and when omitted, the publisher MUST deliver the objects from the most recent group.
+
+* Object Sequence: 
+Identifies the object within the track to start the media delivery. The `Group Sequence` field MUST be set to identify the group under which to
+the deliver the objects. This field is optional and when omitted, the publisher MUST deliver starting from the beginning of the selected group.
```
[the remark this answers is not in the document]
Comment (kixelated) [reply draft-lcurley-warp.md:698]: But can we remove this for now? I don't want to implement it, and I think the usefulness is nebulous at best.

File: draft-lcurley-warp.md [hunk 23]
```diff
@@ -650,30 +652,110 @@ This contains base information required to decode OBJECT messages, such as codec
...
+~~~
+enum Track Result
+{
+  Ok(0),
+  Fail(1)
+}
+
+TRACK RESPONSE {
+  Track Name Length(i),
+  Track Name(...)...,
```
[the remark this answers is not in the document]
Comment (suhasHere) [reply draft-lcurley-warp.md:738]: It correlates the track in the request, since there can be multiple in flight.
we can add further text for handling edge cases or error flows in the follow up PRs.

File: draft-lcurley-warp.md [hunk 24]
```diff
@@ -715,29 +717,89 @@ This contains base information required to decode OBJECT messages, such as codec
...
 
-The format of SUBSCRIBE is as follows:
+The format of SUBSCRIBE REQUEST is as follows:
 
 ~~~
-SUBSCRIBE Message {
-  Track Count (i),
-  Track IDs (..),
+Track Parameter {
+  Track Parameter ID (i),
```
Comment (afrind) [inline draft-lcurley-warp.md:728]: I see these IDs are defined below -- can you add a reference to that section.

File: draft-lcurley-warp.md [hunk 25]
```diff
@@ -784,6 +846,19 @@ The ROLE parameter (key 0x00) allows the client to specify what roles it expects
 
 The client MUST send a ROLE parameter with one of the three values specified above. The server MUST close the connection if the ROLE parameter is missing, is not one of the three above-specified values, or it is different from what the server expects based on the application in question.
 
+# Track Parameters {#track-params}
+
+The Track parameters are described as below. Every parameter MUST appear at most once. The peers SHOULD verify that and close the connection if a parameter appears more than once. The Parameter Value Length field indicates the length of the Parameter Value.
+
+### GROUP SEQUENCE Parameter
+
+The GROUP SEQUENCE parameter (key 0x00) identifies the group within the track to start delivering the objects. The publisher MUST start delivering the objects from the most recent group, when this parameter is omitted.
```
Comment (afrind) [inline draft-lcurley-warp.md:855]: In the case where the publisher is a cache and the media is no longer live, does this mean start from the last group?

File: draft-lcurley-warp.md [hunk 26]
```diff
@@ -784,6 +846,19 @@ The ROLE parameter (key 0x00) allows the client to specify what roles it expects
...
+
+### OBJECT SEQUENCE Parameter
+The OBJECT SEQUENCE parameter (key 0x01) identifies the object with the track to start delivering the media. The `GROUP SEQUENCE` parameter MUST be set to identify the group under which to start the media delivery. The publisher MUST start delivering from the beginning of the selected group when this parameter is omitted.
+
+### AUTHORIZATION INFO Parameter
+AUTHORIZATION INFO parameter (key 0x02) identifies the mandatory parameter carrying track's authorization, authorizing the client’s subscription to the track. The specifics of obtaining the authorization information is out of scope for this specification.
```
Comment (afrind) [inline draft-lcurley-warp.md:861]: If it's mandatory, do you want a MUST in here?  Aren't there applications where authorization isn't required at the track level (either public broadcast, or perhaps authorized at webtransport level for that mapping) - so maybe it shouldn't be mandatory?

File: draft-lcurley-warp.md [hunk 27]
```diff
@@ -715,29 +717,89 @@ This contains base information required to decode OBJECT messages, such as codec
...
-The number of track IDs that follow.
-This MAY be zero to unsubscribe to all tracks.
 
-* Track IDs:
-A list of varint track IDs.
+* Track URI:
+Identifies the track as defined in ({{track-uri}}).
+
+* Track ID: 
+Session specific identifier that maps the Track URI to the Track ID in OBJECT ({{message-object}}) message headers for the advertised track. Peer processing the request message MAY end up choosing a different Track ID (see {{message-subscribe-ok}}). Track IDs are generally shorter than Track URIs and thus reduce the overhead in OBJECT messages.
```
Comment (afrind) [inline draft-lcurley-warp.md:747]: I recommend removing Track ID from SUBSCRIBE REQUEST and leave it only in SUBSCRIBE OK.

**Review by afrind — COMMENTED** [review]
Individual review

File: draft-lcurley-warp.md (hunk 26 above)
Comment (afrind) [reply draft-lcurley-warp.md:861]: Sounds good to me

File: draft-lcurley-warp.md [hunk 28]
```diff
@@ -715,29 +717,89 @@ This contains base information required to decode OBJECT messages, such as codec
...
+
+A `SUBSCRIBE OK` control message is sent for successful subscriptions. 
+
+~~~
+SUBSCRIBE OK
+{
+  Track URI Length(i),
+  Track URI(...),
+  Track ID(i),
+  Expires (i)
```
Comment (vasilvv) [inline draft-lcurley-warp.md:764]: Do we always need `Expires`?  Do we want to have key/value pairs in the OK message, making `Expires` one of those?

File: draft-lcurley-warp.md [hunk 29]
```diff
@@ -715,29 +717,89 @@ This contains base information required to decode OBJECT messages, such as codec
...
+
+Subscriptions stay active until it expires or the underlying transport is disconnected.
+
+
+## SUBSCRIBE ERROR {#message-subscribe-error}
+
+A `SUBSCRIBE ERROR` control message is sent for unsuccessful subscriptions. 
+
+~~~
+SUBSCRIBE ERROR
```
Comment (vasilvv) [inline draft-lcurley-warp.md:787]: Do we want to also add a machine-readable error code?

File: draft-lcurley-warp.md [hunk 30]
```diff
@@ -715,29 +717,89 @@ This contains base information required to decode OBJECT messages, such as codec
...
+Entities that intend to receive media will do so via subscriptions to one or more tracks. The information for the tracks can be obtained via catalog or from incoming SUBSCRIBE REQUEST messages.
 
-The format of SUBSCRIBE is as follows:
+The format of SUBSCRIBE REQUEST is as follows:
 
 ~~~
-SUBSCRIBE Message {
-  Track Count (i),
-  Track IDs (..),
+Track Parameter {
```
[the remark this answers is not in the document]
Comment (suhasHere) [reply draft-lcurley-warp.md:727]: @vasilvv since @afrind suggested we might want to reuse these parameters for Publish Request as well, how about I call it as 'Track Request Parameters'

**Comment by suhasHere** [comment]
@vasilvv @afrind the last few commits should address your feedback. please let me know . Thanks

File: draft-lcurley-warp.md [hunk 31]
```diff
@@ -833,6 +907,7 @@ The producer and consumer MUST cancel a stream, preferably the lowest priority,
 TODO: fill out currently missing registries:
 * Warp version numbers
 * SETUP parameters
+* Track Request parameters
```
Comment (afrind) [inline draft-lcurley-warp.md:910]: +Subscribe Error codes

File: draft-lcurley-warp.md (hunk 25 above)
Comment (afrind) [reply draft-lcurley-warp.md:854]: I can imagine this tripping some people, who would expect that the default is the beginning and not the end.  I suppose it makes sense given the dominant use cases are live.

**Review by afrind — COMMENTED** [review]
Thank you for the updates.

**Review by fluffy — APPROVED** [review]
LGTM

**Review by vasilvv — APPROVED** [review]
My issues:

1. We should have parameters in SUBSCRIBE_OK, since currently it has no extensibility.
2. Expiry needs more thought.

I'll file issues for those, but for now, LGTM.
